vernac/metasyntax.ml

```diff
@@ -665,15 +665,21 @@ let expand_list_rule s typ tkl x n p ll =
        aux (i+1) (main :: tks @ hds) ll in
   aux 0 [] ll
 
-let is_constr_typ typ x etyps =
+let is_constr_typ (s,lev) x etyps =
   match List.assoc x etyps with
-  | ETConstr (_,_,typ') -> typ = typ'
+  (* TODO: factorize these rules with the ones computing the effective
+     sublevel sent to camlp5, so as to include the case of
+     DefaultLevel which are valid *)
+  | ETConstr (s',_,(lev',InternalProd | (NumLevel _ | NextLevel as lev'), _)) ->
+    Notation.notation_entry_eq s s' && lev = lev'
```

Comment:
While you're at it you should probably define the equality function for `production_level`.

Reply:
OK, done.

Somehow, this should be an "attribute" provided by ocaml. There is nothing interesting in mechanically writing a canonical monomorphic equality function. To have it mechanically generated would also prevent possible typos.